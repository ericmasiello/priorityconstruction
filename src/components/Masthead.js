import React, { Component } from 'react';
import ReactDOM from 'react-dom';
import PropTypes from 'prop-types';
import styled from 'styled-components';
import GatsbyImage from './GatsbyImage';
import * as CustomPropTypes from '../propTypes';
import { COLORS } from '../styles/vars';

const BackgroundImage = GatsbyImage.extend`
  left: 0;
  top: 0;
  width: 100%;
  height: 100%;
  opacity: ${props => props.opacity};
`;

export class Masthead extends Component {
  static displayName = 'Masthead';

  static propTypes = {
    tag: PropTypes.oneOfType([PropTypes.func, PropTypes.string]),
    children: PropTypes.node,
    isFullHeight: PropTypes.bool,
    bgImage: CustomPropTypes.ImageSharp.isRequired,
    bgColor: PropTypes.string,
    opacity: PropTypes.number,
  };

  static defaultProps = {
    tag: 'div',
    bgColor: COLORS.bg,
    opacity: 1,
  };

  componentDidUpdate = (prevProps) => {
    const { isFullHeight } = this.props;
    if (isFullHeight !== prevProps.isFullHeight && this.elm && this.elm.animate) {
      if (isFullHeight) {
        this.elm.animate([{ height: '20vh' }, { height: '70vh' }], {
          duration: 300,
          fill: 'forwards',
          easing: 'cubic-bezier(0.86, 0, 0.07, 1)',
          iterations: 1,
        });
      } else {
        this.elm.animate([{ height: '70vh' }, { height: '20vh' }], {
          duration: 300,
          fill: 'forwards',
          easing: 'cubic-bezier(0.86, 0, 0.07, 1)',
          iterations: 1,
        });
      }
    }
  };

  setRef = (elm) => {
    // eslint-disable-next-line react/no-find-dom-node
    this.elm = ReactDOM.findDOMNode(elm);
  };

  render() {
    const {
      bgColor,
      bgImage,
      isFullHeight,
      children,
      opacity,
      tag: Tag,
      ...rest
    } = this.props;

    return (
      <Tag ref={this.setRef} {...rest}>
        {children}
        <BackgroundImage
          style={{
            position: 'absolute',
          }}
          sizes={bgImage.sizes}
          opacity={opacity}
        />
      </Tag>
    );
  }
}

export default styled(Masthead)`
  background-color: ${props => props.bgColor};
  margin-bottom: 1.45rem;
  overflow: hidden;
  position: relative;
  height: ${({ isFullHeight }) => (isFullHeight ? '70vh' : '20vh')};
`;
