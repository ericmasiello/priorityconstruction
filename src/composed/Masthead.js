import React from 'react';
import styled from 'styled-components';
import Masthead from '../components/Masthead';
import HeaderBar from '../components/HeaderBar';
import FlatList from '../components/FlatList';
import Logo from '../components/Logo';
import MainNavLink from '../components/MainNavLink';
import * as CustomPropTypes from '../propTypes';
import { pxToRem } from '../styles/utils';
import LayoutContext from '../layoutContext';

export const ComposedMasthead = (props) => {
  const {
    background,
    ...rest
  } = props;
  return (
    <LayoutContext.Consumer>
      {(layout) => (
        <Masthead
          {...rest}
          bgImage={background}
          isFullHeight={layout.isFullHeight}
        >
          <HeaderBar>
            <Logo image={layout.logo} />
            <nav>
              <FlatList>
                <FlatList.Item>
                  <MainNavLink to="/services">Services</MainNavLink>
                </FlatList.Item>
                <FlatList.Item>
                  <MainNavLink to="/about">About</MainNavLink>
                </FlatList.Item>
                <FlatList.Item>
                  <MainNavLink to="/careers">Careers</MainNavLink>
                </FlatList.Item>
                <FlatList.Item>
                  <MainNavLink to="/contact">Contact</MainNavLink>
                </FlatList.Item>
              </FlatList>
            </nav>
          </HeaderBar>
        </Masthead>
      )}
    </LayoutContext.Consumer>
  );
};

ComposedMasthead.displayName = 'ComposedMasthead';

ComposedMasthead.propTypes = {
  background: CustomPropTypes.ImageSharp.isRequired,
};

export default styled(ComposedMasthead)`
  ${FlatList.Item} {
    margin-right: ${pxToRem(30)};

    &:last-child {
      margin-right: 0;
    }
  }
`;
