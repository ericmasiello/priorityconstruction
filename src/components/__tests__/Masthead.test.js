import React from 'react';
import renderer from 'react-test-renderer';
import { shallow, mount } from 'enzyme';
import 'jest-styled-components';
import Masthead, { Masthead as Raw } from '../Masthead';

const props = {
  bgImage: {
    sizes: {
      src: '/foo/bar.jpg',
    },
  },
};

it('should render', () => {
  const component = renderer.create(<Masthead {...props} />);

  const tree = component.toJSON();
  expect(tree).toMatchSnapshot();
});

it('should render children', () => {
  const wrapper = mount(<Raw {...props}>Test</Raw>);

  expect(wrapper.text()).toContain('Test');
});

it('should render additional props', () => {
  const wrapper = shallow(<Raw {...props} data-test="my-attribute" />);

  expect(wrapper.find('[data-test="my-attribute"]')).toHaveLength(1);
});

it('should display additional classes', () => {
  const wrapper = shallow(<Raw {...props} className="my-class" />);

  expect(wrapper.hasClass('my-class')).toBe(true);
});

it('should render as a <div> by default', () => {
  const wrapper = shallow(<Raw {...props} />);

  expect(wrapper.type()).toBe('div');
});

it('should render with a custom tag', () => {
  const wrapper = shallow(<Raw {...props} tag="span" />);

  expect(wrapper.type()).toBe('span');
});
