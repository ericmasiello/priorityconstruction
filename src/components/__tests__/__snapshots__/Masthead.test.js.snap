// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render 1`] = `
.c1 {
  max-width: 100%;
  left: 0;
  top: 0;
  width: 100%;
  height: 100%;
  opacity: 1;
}

.c0 {
  margin-bottom: 1.45rem;
  overflow: hidden;
  position: relative;
  height: 20vh;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  max-height: 43.75rem;
}

.c0 .sc-htpNat {
  position: relative;
  z-index: 2;
  color: #FBDB1E;
  background-color: rgba(0,0,0,0.5);
  text-align: center;
  width: 100%;
  padding-top: 1rem;
  padding-bottom: 1rem;
  margin-bottom: 0;
}

<div
  className="c0"
>
  <div
    className=" gatsby-image-outer-wrapper"
    style={
      Object {
        "position": "initial",
      }
    }
  >
    <div
      className="c1 gatsby-image-wrapper"
      style={
        Object {
          "overflow": "hidden",
          "position": "absolute",
        }
      }
    >
      <div
        style={
          Object {
            "paddingBottom": "NaN%",
            "width": "100%",
          }
        }
      />
      <img
        alt=""
        onLoad={[Function]}
        src="/foo/bar.jpg"
        style={
          Object {
            "height": "100%",
            "left": 0,
            "objectFit": "cover",
            "objectPosition": "center",
            "opacity": 1,
            "position": "absolute",
            "top": 0,
            "transition": "opacity 0.5s",
            "width": "100%",
          }
        }
      />
      <noscript
        dangerouslySetInnerHTML={
          Object {
            "__html": "<img src=\\"/foo/bar.jpg\\" alt=\\"\\" style=\\"position:absolute;top:0;left:0;transition:opacity 0.5s;transition-delay:0.5s;opacity:1;width:100%;height:100%;object-fit:cover;object-position:center\\"/>",
          }
        }
      />
    </div>
  </div>
</div>
`;
